### PR TITLE
add UPnP IGD port forwarding support

### PR DESCRIPTION
**Pull request by vinipsmaker**
Related to [MAID-1132](https://maidsafe.atlassian.net/browse/MAID-1132).

[Review on Reviewable](https://reviewable.io/reviews/maidsafe/crust/181)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @chandraprakash

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/tcp_connections.rs [hunk 1]
```diff
@@ -76,6 +78,36 @@ pub fn listen(port: u16) -> IoResult<(Receiver<(TcpStream, SocketAddr)>, TcpList
...
+                                }
+                            });
+                        },
+                        _ => (),
+                    }
+                },
+                Err(_) => (),
+            }
+        }
         loop {
```
Comment (chandraprakash) [inline src/tcp_connections.rs]: Nice work ! @vinipsmaker 
I understand your thoughts regarding the design here, but adding port mapping utility to *transport.rs* makes it protocol agnostic. That means we can have port mapping facility for any new protocol (if supported) automatically in place.

File: src/tcp_connections.rs [hunk 2]
```diff
@@ -76,6 +78,36 @@ pub fn listen(port: u16) -> IoResult<(Receiver<(TcpStream, SocketAddr)>, TcpList
...
+                            let _ = thread::Builder::new()
+                                .name("IGD".to_string()).spawn(move || {
+                                match igd::search_gateway_from(ifa) {
+                                    Ok(gateway) => {
+                                        let _ = gateway
+                                            .add_port(igd::PortMappingProtocol::TCP,
+                                                      port,
+                                                      SocketAddrV4::new(ifa, port),
+                                                      0, "crust");
+                                    },
```
Comment (chandraprakash) [inline src/tcp_connections.rs]: we are missing ability to know our_external_endpoint (call to `gateway::get_external_ip()`). This information needs to be stored in connection manager and should be provided on call to  `connection_manager::get_own_endpoints() -> Vec<Endpoint>` (Returning external endpoint (if IGD Protocol supported) along with all its own listening endpoint (all interfaces)).

**Comment by chandraprakash** [comment]
Crust README.md documentation (pre-requisite section) needs to be updated with additional dependency  IGD -> hyper 0.6.* - > openssl ^0.6 . 
rust-openssl depends on both the OpenSSL runtime libraries and headers.
https://github.com/sfackler/rust-openssl

see https://github.com/maidsafe/routing#pre-requisite for example

**Comment by SBSTP** [comment]
Hey! I'm the `igd` crate maintainer. I don't think we actually need SSL to talk with the router. Hyper offers ssl under a feature, so I'm going to disable that feature. So you won't need OpenSSL anymore.

**Comment by chandraprakash** [comment]
Thanks a lot @SBSTP !!! You have been very helpful in reviewing PRs and publishing igd library.

**Comment by dirvine** [comment]
+1 @chandraprakash  I agree @SBSTP thanks very much indeed. Got to love all the collaboration

**Comment by vinipsmaker** [comment]
<sup>**[src/tcp_connections.rs, line 100 \[r1\]](https://reviewable.io:443/reviews/maidsafe/crust/181#-JtlTwsQl6G0K35_XYa2-r1-100)** ([raw file](https://github.com/maidsafe/crust/blob/a4344f3fb12d3eb578fcc28911d62618ca0ff7b0/src/tcp_connections.rs#L100)):</sup>
I pushed a new commit with changes related to this request. Please review if you like how the "async" nature is handled or if I should use other communication mechanism among the threads.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/crust/181)

File: src/connection_manager.rs [hunk 3]
```diff
@@ -448,6 +499,17 @@ impl ConnectionManager {
...
+    /// Return external endpoints. That is, the endpoints other peers can use to
+    /// connect to. If an external IP is not available, the IP obtained through
+    /// the OS is used. The external IP is obtained through UPnP IGD.
+    pub fn get_own_endpoints(&self) -> Vec<Endpoint> {
+        self.own_endpoints.iter()
+            .map(|&(ref local, ref external)| match *external.lock().unwrap() {
+                Some(ref a) => a,
+                None => local,
+            }.clone()).collect()
+    }
```
Comment (chandraprakash) [inline src/connection_manager.rs]: I would return all endpoints here (internal and external(if available)).

**Comment by vinipsmaker** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[src/connection_manager.rs, line 512 \[r3\]](https://reviewable.io:443/reviews/maidsafe/crust/181#-JtyyI3XAukpV3IDFq1l-r3-512)** ([raw file](https://github.com/maidsafe/crust/blob/dc8cbfe0cbc1b090d667669e1df1edf2f03e600b/src/connection_manager.rs#L512)):</sup>
I pushed a new commit with this change. Please review it.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/crust/181)

**Comment by chandraprakash** [comment]
Reviewed 1 of 3 files at r1, 1 of 1 files at r2, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/crust/181)

**Comment by chandraprakash** [comment]
Thanks @vinipsmaker  !
Review done. Please update the PR and ensure all the tests are passing.

**Comment by vinipsmaker** [comment]
> Review done. Please update the PR and ensure all the tests are passing.

Done.

**Comment by chandraprakash** [comment]
@Fraser999  can you please help with  jenkins-ci-win64 build.

**Comment by Fraser999** [comment]
@chandraprakash The OS X CI machine was hung running another Crust test.  I've killed that one which should allow this to run.  The Windows tests will fail until non-MaidSafe dependencies are fixed to work with the latest nightly build of Rust.

**Comment by Fraser999** [comment]
And this test hung too on the Mac, so I've killed it.

**Comment by chandraprakash** [comment]
@vinipsmaker  Can you look into  jenkins-ci-osx failure please.
@Fraser999  can help you to get access to one of the mac.(if required)

**Comment by vinipsmaker** [comment]
Yes, I'd like to have more info about the environment.

```
test connection_manager::test::connection_manager ... Build was aborted
Process didn't exit successfully: `/Users/jenkins/workspace/crust_osx/target/debug/crust-d25869973e168d1b` (signal: 15)
```

**Comment by vinipsmaker** [comment]
Are you sure `master` is building on OS X?

I get the same errors with my commit:

```
failures:

---- tcp_connections::test::test_multiple_nodes_small_stream stdout ----
        thread 'tcp_connections::test::test_multiple_nodes_small_stream' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 24, message: "Too many open files" } }', ../src/libcore/result.rs:731



failures:
    tcp_connections::test::test_multiple_nodes_small_stream
```

And without my commit:

```
---- tcp_connections::test::test_multiple_nodes_small_stream stdout ----
        thread 'tcp_connections::test::test_multiple_nodes_small_stream' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 24, message: "Too many open files" } }', ../src/libcore/result.rs:731



failures:
    tcp_connections::test::test_multiple_nodes_small_stream
```

**Comment by vinipsmaker** [comment]
I can do a git-bisect if you want me to.

**Comment by dirvine** [comment]
Are you running in a single thread, some tests fail if you allow them to run in parallel. This may be one.

**Comment by vinipsmaker** [comment]
> Are you running in a single thread

Yes, single-threaded. I've have set the env var `RUST_TEST_THREADS=1` before `cargo test`.

**Comment by chandraprakash** [comment]
master builds ok on mac (Jenkins - http://ci.maidsafe.net:8080/job/crust_osx_status_badge/lastBuild/console)

**Comment by vinipsmaker** [comment]
First time I debug something on MAC/lldb.

Anyway, how do I manage to make the `master` branch succeed on the tests? [It's failing in this line](https://github.com/maidsafe/crust/blob/2c37e43d5b62fd2a39212b9396c8f7711d77cd6b/src/tcp_connections.rs#L206). The second wrap (`Result<(std::sync::mpsc::Receiver<(u64, u64)>, crust::tcp_connections::OutTcpStream<u64>), std::io::error::Error>`) fails with `Error { repr: Os { code: 24, message: "Too many open files" } }`. It's completely unrelated to the changes 

I'm a full n00b in Mac and I don't know how to raise the limits:

```
prakashs-iMac:crust vinipsmaker$ ulimit
unlimited
```

**Comment by vinipsmaker** [comment]
@chandraprakash, [build is fine on OS X](http://ci.maidsafe.net:8080/job/crust_osx/148/).

**Comment by dirvine** [comment]
Yes Mac has very low ulimit set so we need ot restrict tests running too many handles (or the machine doing other stuff).

**Pull request merged**